tensorboard/plugins/graph/tf_graph_common/parser.ts

```diff
@@ -109,50 +109,68 @@ export function fetchAndParseGraphData(path: string, pbTxtFile: Blob,
  * @param callback The callback called on each line
  * @param chunkSize The size of each read chunk. (optional)
  * @param delim The delimiter used to split a line. (optional)
- * @returns A promise for when it is finished.
+ * @returns Promise that resolves with true when it is finished.
  */
 export function streamParse(
     arrayBuffer: ArrayBuffer, callback: (string) => void,
     chunkSize: number = 1000000, delim: string = '\n'): Promise<boolean> {
   return new Promise<boolean>(function(resolve, reject) {
     let offset = 0;
-    let bufferSize = arrayBuffer.byteLength - 1;
     let data = '';
 
-    function readHandler(str) {
+    function readHandler(str: string) {
       offset += chunkSize;
       let parts = str.split(delim);
       let first = data + parts[0];
       if (parts.length === 1) {
         data = first;
-        readChunk(offset, chunkSize);
+        readChunkIfContent(offset, chunkSize);
         return;
       }
       data = parts[parts.length - 1];
-      callback(first);
-      for (let i = 1; i < parts.length - 1; i++) {
-        callback(parts[i]);
-      }
-      if (offset >= bufferSize) {
-        if (data) {
-          callback(data);
+      try {
+        callback(first);
+        for (let i = 1; i < parts.length - 1; i++) {
+          callback(parts[i]);
         }
-        resolve(true);
+      } catch (e) {
+        reject(e);
         return;
       }
-      readChunk(offset, chunkSize);
+      readChunkIfContent(offset, chunkSize);
     }
 
-    function readChunk(offset: number, size: number) {
+    /**
+     * Reads a chunk of data from the ArrayBuffer starting from offset.
+     *
+     * It terminates (resolves the promise) if there is no more content to be
+     * read.
+     * @param offset Starting index of content to read from the Buffer.
+     * @param size The size of each read chunk. (optional)
+     */
+    function readChunkIfContent(offset: number, size: number) {
```

Comment:
The code here feels a lot more complicated than necessary with two different functions that each invoke the other, and the callback being invoked in three different places.  Also the chunk size doesn't need to be passed around since it's a constant within any invocation of `streamParse()`.  Could we simplify to something like this:

```ts
function readChunk(oldData, newData, offset) {
  let doneReading = offset >= arrayBuffer.bytelength;
  let parts = newData.split(delim);
  parts[0] = oldData + parts[0];
  let remainder = '';
  if (!doneReading) {
    remainder = parts.pop();
  }
  for (let part of parts) {
    try {
      callback(part);
    } catch (e) {
      reject(e):
      return;
    }
  }
  if (doneReading) {
    resolve(true);
    return;
  }
  const nextChunk = new Blob([arrayBuffer.slice(offset, offset + chunkSize)]);
  const file = new FileReader();
  file.onload = (e: any) => readChunk(remainder, e.target.result, offset + chunkSize);
  file.readAsText(nextChunk);
}

readChunk('', '', 0);
```

Reply:
Good idea! The code was hard to read as is. The new code is an improvement!